blockchain/chain_sync/src/chain_muxer.rs

```diff
@@ -0,0 +1,940 @@
+#[cfg(test)]
+mod peer_test;
+
+use crate::bad_block_cache::BadBlockCache;
+use crate::network_context::HelloResponseFuture;
+use crate::network_context::SyncNetworkContext;
+use crate::sync_state::SyncState;
+use crate::tipset_syncer::{
+    TipsetProcessor, TipsetProcessorError, TipsetRangeSyncer, TipsetRangeSyncerError,
+};
+use crate::validation::{TipsetValidationError, TipsetValidator};
+
+use amt::Amt;
+use beacon::{Beacon, BeaconSchedule};
+use blocks::{
+    Block, Error as ForestBlockError, FullTipset, GossipBlock, Tipset, TipsetKeys, TxMeta,
+};
+use chain::{ChainStore, Error as ChainStoreError};
+use cid::{Cid, Code::Blake2b256};
+use encoding::{Cbor, Error as EncodingError};
+use fil_types::verifier::ProofVerifier;
+use forest_libp2p::{
+    hello::HelloRequest, rpc::RequestResponseError, NetworkEvent, NetworkMessage, PubsubMessage,
+};
+use ipld_blockstore::BlockStore;
+use libp2p::core::PeerId;
+use message::{SignedMessage, UnsignedMessage};
+use message_pool::{MessagePool, Provider};
+use networks::BLOCK_DELAY_SECS;
+use state_manager::StateManager;
+
+use async_std::channel::{Receiver, Sender};
+use async_std::pin::Pin;
+use async_std::stream::StreamExt;
+use async_std::sync::{Mutex, RwLock};
+use async_std::task::{self, Context, JoinHandle, Poll};
+use futures::{
+    future,
+    future::try_join_all,
+    future::{Future, FutureExt},
+    stream::TryStreamExt,
+    try_join, Stream,
+};
+use futures::{select, stream::FuturesUnordered};
+use log::{debug, error, info, trace, warn};
+use serde::Deserialize;
+use thiserror::Error;
+
+use std::sync::Arc;
+use std::{
+    marker::PhantomData,
+    time::{SystemTime, UNIX_EPOCH},
+};
+
+const MAX_HEIGHT_DRIFT: u64 = 5;
+
+// TODO revisit this type, necessary for two sets of Arc<Mutex<>> because each state is
+// on separate thread and needs to be mutated independently, but the vec needs to be read
+// on the RPC API thread and mutated on this thread.
+type WorkerState = Arc<RwLock<Vec<Arc<RwLock<SyncState>>>>>;
```

Comment:
Probably can remove the outer most `Arc<RwLock<Vec>>` given the new architecture